### PR TITLE
Fixed popup positioning issue

### PR DESCRIPTION
**Pull request by YohDeadfall**
Fixes popup positioning issue caused by #5462. `HasFlag` and `HasFlagCustom` don't check that all specified bits are set which isn't how masked values work. Thefeore, I renamed `HasFlagCustom` to `HasAllFlags` and added `HasAnyFlag` which returns `true` if any of the specified bits is set.

### CODE REVIEW

File: src/Avalonia.Controls/Primitives/PopupPositioning/IPopupPositioner.cs [hunk 1]
```diff
@@ -265,25 +265,25 @@ public static void ValidateGravity(this PopupGravity gravity)
 
         public static PopupAnchor Flip(this PopupAnchor edge)
         {
-            if (edge.HasFlagCustom(PopupAnchor.HorizontalMask))
+            if (edge.HasAnyFlag(PopupAnchor.HorizontalMask))
```
Comment (YohDeadfall) [inline src/Avalonia.Controls/Primitives/PopupPositioning/IPopupPositioner.cs:268]: Here and below are lines which cause the positioning problem in the `master` branch.

**Comment by grokys** [comment]
`HasFlagCustom` was a public API so this is a breaking change.

**Comment by Gillibald** [comment]
I don't see why we would need `HasFlagCustom` if it doesn't check for set bits. `HasFlagCustom` needs to be fixed. We don't need a new helper that does what we expect from `HasFlagCustom`

**Comment by YohDeadfall** [comment]
`HasAllFlags` does `(value & flag) == flag` and it's the same logic `HasFlag` and `HasFlagCustom` do.
`HasAnyFlag` does `(value & flag) != 0` which is completely different thing.

**Comment by Gillibald** [comment]
I don't see a use case for `(value & flag) == flag`
Change the current implementation to `(value & flag) != 0` and if there is a real use case for the exact match introduce a new extension for that.

**Comment by YohDeadfall** [comment]
Thank you for pointing this out, added dependency as requested😉

**Comment by kekekeks** [comment]
@Gillibald Enum.HasFlag has exactly the same behavior as our current HasFlagCustom - it checks if ALL bits are present.

We should keep the current `HasFlagCustom` as is and either introduce `MatchesFlagMask` or just remove HasFlagCustom from the relevant parts of the popup positioner.

**Comment by Gillibald** [comment]
Then the naming is wrong but let's not go too off-topic

**Comment by Gillibald** [comment]
We have multiple `HasFlagCustom` usages here:https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Primitives/PopupPositioning/IPopupPositioner.cs#L256

Please help me understand the issue here. Do we successfully validate a combination of set bits that shouldn't be valid? So we should test against the combination of both flags instead?

**Comment by YohDeadfall** [comment]
The code you pointed out is a good example of when `HasFlag` should be used. If all bits are set (`Right` and `Left`) then it's not a valid value. That's why the current behavior should be kept. By the way, I optimized that code already by 615179a689b63f178abba6120c7d1d42b2a131cd.

`HasAnyFlag` is useful in a different situation, when you need to check that at least one bit is present of the specified set.

**Review by grokys — APPROVED** [review]

**Pull request merged**